lib/routes/password.js

```diff
@@ -433,85 +433,79 @@ module.exports = function (
-              // Clobber the timestamp to prevent prematurely expired tokens.
-              accountRecord.createdAt = undefined
-              return db.createPasswordForgotToken(accountRecord)
-            }
-          )
-          .then(
-            function (passwordForgotToken) {
-              return db.accountEmails(passwordForgotToken.uid)
-                .then(emails => {
-                  const geoData = request.app.geo
-                  const {
-                    browser: uaBrowser,
-                    browserVersion: uaBrowserVersion,
-                    os: uaOS,
-                    osVersion: uaOSVersion,
-                    deviceType: uaDeviceType
-                  } = request.app.ua
-
-                  return mailer.sendRecoveryCode(emails, passwordForgotToken, {
-                    token: passwordForgotToken,
-                    code: passwordForgotToken.passCode,
-                    service: service,
-                    redirectTo: request.payload.redirectTo,
-                    resume: request.payload.resume,
-                    acceptLanguage: request.app.acceptLanguage,
-                    flowId: flowId,
-                    flowBeginTime: flowBeginTime,
-                    ip: ip,
-                    location: geoData.location,
-                    timeZone: geoData.timeZone,
-                    uaBrowser,
-                    uaBrowserVersion,
-                    uaOS,
-                    uaOSVersion,
-                    uaDeviceType,
-                    uid: passwordForgotToken.uid
-                  })
-                })
-                .then(
-                  function () {
-                    return request.emitMetricsEvent('password.forgot.send_code.completed')
-                  }
-                )
-                .then(
-                  function () {
-                    return passwordForgotToken
-                  }
-                )
+          .then(accountRecord => {
+            if (accountRecord.primaryEmail.normalizedEmail !== email.toLowerCase()) {
+              throw error.cannotResetPasswordWithSecondaryEmail()
             }
-          )
-          .then(
-            function (passwordForgotToken) {
-              return {
-                passwordForgotToken: passwordForgotToken.data,
-                ttl: passwordForgotToken.ttl(),
-                codeLength: passwordForgotToken.passCode.length,
-                tries: passwordForgotToken.tries
-              }
-            },
-
-          )
+            // The token constructor sets createdAt from its argument.
+            // Clobber the timestamp to prevent prematurely expired tokens.
+            accountRecord.createdAt = undefined
+            return db.createPasswordForgotToken(accountRecord)
+          })
+          .then(result => {
+            passwordForgotToken = result
+            return request.stashMetricsContext(passwordForgotToken)
```

Comment:
I imagine that the `request.stashMetricsContext` is a fairly fast method? No significant benefit to group in `Promise.resolve` with `db.accountEmails`?

Reply:
Good point, may as well do them concurrently.